File: apps/desktop/src/services/server/relationships.ts
import type {
  RelationshipPage,
  ViewerRelationship,
  RelationshipCreateRequest,
  RelationshipChangeRequest,
  KnowledgeMaintenanceRequest,
  GraphStatus,
} from "@meowlive/contracts";
import {
  adminTransport,
  record,
  strings,
  integer,
  viewerPath,
  type ClientOptions,
} from "./companionship";
const entity = (v: unknown) => record(v) && strings(v, ["kind", "id"]);
const fact = (v: unknown) =>
  record(v) &&
  strings(v, ["id", "kind", "confirmation"]) &&
  integer(v.version) &&
  entity(v.source) &&
  entity(v.target) &&
  typeof v.deleted === "boolean" &&
  (v.expires_at_ms === null || integer(v.expires_at_ms)) &&
  Array.isArray(v.evidence) &&
  v.evidence.every(
    (e) => record(e) && strings(e, ["source", "event_id", "quote"]),
  );
export function createRelationshipClient(options: ClientOptions = {}) {
  const req = adminTransport(options);
  return {
    list: (id: string, depth: number, signal?: AbortSignal) =>
      req<RelationshipPage>(
        `${viewerPath(id)}/relationships?depth=${depth}`,
        (v) =>
          record(v) &&
          typeof v.degraded === "boolean" &&
          Array.isArray(v.relationships) &&
          v.relationships.every(fact),
        undefined,
        signal,
      ),
    create: (r: RelationshipCreateRequest, signal?: AbortSignal) =>
      req<ViewerRelationship>("/api/admin/relationships", fact, r, signal),
    change: (id: string, r: RelationshipChangeRequest, signal?: AbortSignal) =>
      req<ViewerRelationship>(
        `/api/admin/relationships/${encodeURIComponent(id)}`,
        fact,
        r,
        signal,
      ),
    status: (signal?: AbortSignal) =>
      req<GraphStatus>(
        "/api/admin/graph/status",
        (v) =>
          record(v) &&
          ["pending", "leased", "failed"].every((k) => integer(v[k])) &&
          (v.oldest_pending_age_ms === null ||
            integer(v.oldest_pending_age_ms)) &&
          typeof v.connected === "boolean",
        undefined,
        signal,
      ),
    rebuild: (r: KnowledgeMaintenanceRequest, signal?: AbortSignal) =>
      req<{ queued: number }>(
        "/api/admin/graph/rebuild",
        (v) => record(v) && integer(v.queued),
        r,
        signal,
      ),
  };
}
export type RelationshipClient = ReturnType<typeof createRelationshipClient>;

File: crates/protocol/src/relationships.rs
//! 受保护关系事实、证据和图同步状态契约。
use serde::{Deserialize, Serialize};
use ts_rs::TS;
#[derive(Clone, Debug, Deserialize, Serialize, TS)]
#[serde(deny_unknown_fields)]
pub struct RelationshipEntity {
    pub kind: String,
    pub id: String,
}
#[derive(Clone, Debug, Deserialize, Serialize, TS)]
#[serde(deny_unknown_fields)]
pub struct RelationshipEvidence {
    pub source: String,
    pub event_id: String,
    pub quote: String,
}
#[derive(Clone, Debug, Deserialize, Serialize, TS)]
pub struct ViewerRelationship {
    pub id: String,
    #[ts(type = "number")]
    pub version: u64,
    pub source: RelationshipEntity,
    pub target: RelationshipEntity,
    pub kind: String,
    pub confirmation: String,
    pub evidence: Vec<RelationshipEvidence>,
    #[ts(type = "number | null")]
    pub expires_at_ms: Option<u64>,
    pub deleted: bool,
}
#[derive(Clone, Debug, Deserialize, Serialize, TS)]
pub struct RelationshipPage {
    pub relationships: Vec<ViewerRelationship>,
    pub degraded: bool,
}
#[derive(Clone, Debug, Deserialize, Serialize, TS)]
#[serde(deny_unknown_fields)]
pub struct RelationshipCreateRequest {
    pub source: RelationshipEntity,
    pub target: RelationshipEntity,
    pub kind: String,
    pub evidence: Vec<RelationshipEvidence>,
    #[ts(type = "number | null")]
    pub expires_at_ms: Option<u64>,
    pub admin_confirmed: bool,
    pub reason: String,
    pub request_key: String,
}
#[derive(Clone, Debug, Deserialize, Serialize, TS)]
#[serde(deny_unknown_fields)]
pub struct RelationshipChangeRequest {
    #[ts(type = "number")]
    pub expected_version: u64,
    pub action: String,
    pub reason: String,
    pub request_key: String,
}
#[derive(Clone, Debug, Deserialize, Serialize, TS)]
#[serde(deny_unknown_fields)]
pub struct KnowledgeMaintenanceRequest {
    pub request_key: String,
    pub reason: String,
}
#[derive(Clone, Debug, Deserialize, Serialize, TS)]
pub struct GraphStatus {
    #[ts(type = "number")]
    pub pending: u64,
    #[ts(type = "number")]
    pub leased: u64,
    #[ts(type = "number")]
    pub failed: u64,
    #[ts(type = "number | null")]
    pub oldest_pending_age_ms: Option<u64>,
    pub connected: bool,
}
